pkg/importer/dataStream.go

```diff
@@ -734,31 +920,51 @@ func (d *DataStream) convertQcow2ToRaw(src, dest string) error {
 }
 
 // Copy endpoint to dest based on passed-in reader.
-func (d *DataStream) copy(dest string) error {
+func (d *DataStream) copy(dest string) (err error) {
 	if d.isHTTPQcow2() {
-		err := d.convertQcow2ToRawStream(dest)
+		err = d.convertQcow2ToRawStream(dest)
 		if err != nil {
 			return err
 		}
 	} else {
 		dest = filepath.Clean(dest)
 		glog.V(2).Infof("copying image file to %q", dest)
 		tmpDest := dest
-		if d.qemu {
-			// copy to tmp; qemu conversion will write to passed-in destination
-			tmpDest = randTmpName(dest)
-			glog.V(3).Infof("Copy: temp file for qcow2 conversion: %q", tmpDest)
-			defer func(f string) {
-				os.Remove(f)
-			}(tmpDest)
+		if d.isCopyToTmpRequired() {
+			if d.isQemu() {
+				// copy to tmp; qemu conversion will write to passed-in destination
+				tmpDest = randTmpName(dest)
+				glog.V(3).Infof("Copy: temp file for qcow2 conversion: %q", tmpDest)
+				defer func(f string) {
+					os.Remove(f)
+				}(tmpDest)
+			}
+			if d.isMovble() {
+				//before moving has to close all readers
+				closeReaders(d.Readers)
+				err = MoveFile(d.getTmpDataPath(), tmpDest)
+				if err != nil {
+					//Failback to copy since move failed
```

Comment:
I what situations would this code succeed if a move failed?

Reply:
If source and dest belong to different devices/partitions (tmpPVC vs PVC) Rename will fail with the following error "invalid cross-device link". Copy however, does succeed, since it will actually create a new file on the destination file system

Reply:
oh, got it